ext/atomic/index.md

```diff
@@ -0,0 +1,510 @@
+# Atomic Operations Extension
+
+This extension provides a means to perform multiple "operations" in a linear and
+atomic manner. Operations are a serialized form of the mutations allowed in the
+base JSON:API specification.
+
+Clients can send an array of operations in a single request. This extension
+guarantees that those operations will be processed in order and will either
+completely succeed or fail together.
+
+## <a href="#namespace" id="namespace" class="headerlink"></a> Namespace
+
+This extension uses the namespace `atomic`.
+
+> Note: JSON:API extensions can only introduce new document members using a
+> reserved namespace as a prefix.
+
+## <a href="#document-structure" id="document-structure" class="headerlink"></a> Document Structure
+
+A document that supports this extension **MAY** include any of the top-level
+members allowed by the base specification with the exception of `data` and
+`included`, which **MUST NOT** be included.
+
+In addition, such a document **MAY** include either of the following members,
+but not both:
+
+- `atomic:operations` - an array of [operation objects](#operation-objects).
+
+- `atomic:results` - an array of [result objects](#result-objects).
+
+If either `atomic:operations` or `atomic:results` is present, the `errors`
+member **MUST NOT** be included in the same document.
+
+### <a href="#operation-objects" id="operation-objects" class="headerlink"></a> Operation Objects
+
+An operation object **MUST** contain the following member:
+
+- `op`: an operation code, expressed as a string, that indicates the type of
+  operation to perform. Valid values include:
+
+  - `"add"`: creates a new resource or relationship
+  - `"update"`: updates a resource or relationship
+  - `"remove"`: removes a resource or relationship
+
+An operation object **MAY** contain either of the following members, but not
+both, to specify the target of the operation:
+
+- `ref`: an object that **MAY** contain any of the following combinations of
+  members:
+
+  - `type` and `id`: to target an individual resource.
```

Comment:
At first sight, I think it may induce that both `type` & `id` were required, but of course it's not in case of an "add" operation. Should we be more specific ?
```suggestion
  - `type` and `id`: to target an individual resource. `id` **MAY** be omitted in case of an "add" operation.
```

Reply:
I think I agree with this refinement. What do you think @gabesullice ?

Reply:
Sorry, I misunderstand the context here. After talking with @gabesullice we realized that `ref` is unnecessary in an `add` operation, which will already contain a `type` in the `data` member that represents the resource to be added.

Reply:
Don't we still need a `ref` to represents the "endpoint" to which the resource will be sent ? 
If `ref` is considered unnecessary for `add` (we infer the endpoint from the resource itself), it  also become unnecessary for `update`, right ?

From an implementer POV, I like the idea of just having to look at `ref` (or `href`) to determine to which endpoint pass the data which is treated like an opaque blob at this point.

Reply:
I mean, in this particular context of Atomic Operations, if `ref` is not necessary in an `add` operation, why should we have different endpoints when adding new resources through "plain" JSON:API calls ?  A single /jsonapi/add would suffice.
But (in Drupal JSON:API implementation, at least), `add` operations are performed on different endpoints, based on the resource type, and then there's a check that the data's `type` (and also `id`, for in case of updates) matches the endpoint URL.
I agree that this can be perceived as a kind of "redundancy", but this is already the case with existing "regular" JSON:API, right ?
Let's say I build a JSON:API implementation where *every* write calls goes to a `/jsonapi` accepting PATCH or POST requests and inferring everything from the passed data, would it still be compliant with the original JSON:API spec ?

Reply:
> Let's say I build a JSON:API implementation where every write calls goes to a /jsonapi accepting PATCH or POST requests and inferring everything from the passed data, would it still be compliant with the original JSON:API spec?

Yes, I think it would be. Imagine something like `calendar.mydomain.com/`. Here, I could `POST` reminders and events. A `GET` request to `/` (root) might return a mixed collection of reminders and events in chronological order. Is it good API design? I think you could argue both yes and no, but the spec doesn't prevent it.

---

One of the use cases that I wanted to support by using the `href` member in an operation object was the concept of creating resource objects via editorialized endpoints. For example: adding both a `tshirt` and a `sandal` resource object to a `/summer-styles` endpoint.

I think that's sort of similar to what you're asking. You might be implementing a `/operations` endpoint where you want to add both`tshirt` and `sandal` resource objects:

```json
[{
  "op": "add",
  "ref": {
    "type": "tshirt"
  },
  "data": {
    "type": "tshirt",
    "attributes": {
      "sex": "male",
      "color": "blue"
    }
  }
}, {
  "op": "add",
  "ref": {
    "type": "sandal"
  },
  "data": {
    "type": "sandal",
    "attributes": {
      "sex": "unisex",
      "size": "gargantuan"
    }
  }
}]
```

That feels very redundant to me, honestly. However, you do say:

> From an implementer POV, I like the idea of just having to look at ref (or href) to determine to which endpoint pass the data which is treated like an opaque blob at this point.

Knowing that you're implementing this within Drupal I'll take some liberties with my examples...

I imagine you're talking about routing subrequests and I see why you'd want to treat the `data` member as an opaque blob. That said, you have to deserialize the entire JSON document anyway. It seems like the code below wouldn't be too onerous since you'll have access to the operation data:

```php
  $operation = $deserialized['atomic:operations'][0];
  $url = $operation['href'] ?? Url::fromRoute("jsonapi.{$operation['data']['type']}.collection.post")->toString();
  $subrequest = Request::create($url);
```